### PR TITLE
tree2: Add enum example

### PR DESCRIPTION
**Pull request by CraigMacomber**
## Description

Add example of unions like enums.

Future work could polish up this utility and we could export it. For now this shows how it could be done.

## Reviewer Guidance

The review process is outlined on [this wiki page](https://github.com/microsoft/FluidFramework/wiki/PR-Guidelines#guidelines).

### CODE REVIEW

File: experimental/dds/tree2/src/class-tree/schemaCreationUtilities.ts [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+import { SchemaFactory } from "./schemaFactory";
+// This type is referred to by the base type generated by schema.
+// Adding the import of it here prevents TypeScript from generating inline includes for it in the d.ts file.
+// If we let TypeScript generate these includes, they use relative paths which break API extractor's rollup.
+// eslint-disable-next-line unused-imports/no-unused-imports
+import { type NodeKind as _dummy } from "./schemaTypes";
+
+/**
+ * Create a schema for a node with no state.
+ * @remarks
```
Comment (Josmithr) [inline experimental/dds/tree2/src/class-tree/schemaCreationUtilities.ts:18]: Great docs! Thanks!!

File: experimental/dds/tree2/src/class-tree/schemaCreationUtilities.ts [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+			return name;
+		}
+	};
+}
+
+/**
+ * Converts an enum into a collection of schema which can be used in a union.
+ * @alpha
+ */
+export function adaptEnum<TScope extends string, const TEnum extends Record<string, string>>(
```
Comment (Josmithr) [inline experimental/dds/tree2/src/class-tree/schemaCreationUtilities.ts:41]: This only works for string-based enums, but presumably it wouldn't be too hard to extend this to capture number-based enums, right?

File: experimental/dds/tree2/src/class-tree/schemaCreationUtilities.ts [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+	}
+
+	return out;
+}
+
+/**
+ * `Object.values`, but with more specific types.
+ * @alpha
+ */
+export function typedObjectValues<TKey extends string, TValues>(
```
Comment (Josmithr) [inline experimental/dds/tree2/src/class-tree/schemaCreationUtilities.ts:83]: Does this need to be exported?

File: experimental/dds/tree2/src/class-tree/schemaCreationUtilities.ts [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+	if (names.size !== members.length) {
+		throw new UsageError("All members of enums must have distinct names");
+	}
+
+	type TOut = Record<Members, ReturnType<typeof singletonSchema<TScope, Members>>>;
+	const factoryOut = <TValue extends Members>(value: TValue) => {
+		return new out[value]({}) as unknown as ReturnType<typeof singletonSchema<TScope, TValue>>;
+	};
+	const out = factoryOut as typeof factoryOut & TOut;
+	// const out: Record<Members, ReturnType<typeof makeSchema<Members>>> = Object.create(null);
```
Comment (Josmithr) [inline experimental/dds/tree2/src/class-tree/schemaCreationUtilities.ts:112]: Should be removed?

File: experimental/dds/tree2/src/class-tree/schemaCreationUtilities.ts [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+			value: singletonSchema(factory, name),
+		});
+	}
+
+	return out;
+}
+
+// TODO: Why does this one generate an invalid d.ts file if exported?
+// Tracked by https://github.com/microsoft/TypeScript/issues/56718
+function _enumFromStrings2<TScope extends string, const Members extends readonly string[]>(
```
Comment (Josmithr) [inline experimental/dds/tree2/src/class-tree/schemaCreationUtilities.ts:127]: Does this exist here _only_ to repro the above issue? Doesn't appear to be used within the module.

File: experimental/dds/tree2/src/test/tsconfig.json [hunk 6]
```diff
@@ -4,11 +4,12 @@
 		"lib": ["ES2019", "ES2018.Promise", "DOM", "DOM.Iterable"],
 		"noImplicitAny": true,
 		"noUnusedLocals": false,
+		"noImplicitOverride": true,
 		"rootDir": "./",
 		"outDir": "../../dist/test",
 		"types": ["node", "mocha"],
 	},
-	"include": ["./**/*"],
+	"include": ["./**/*", "../class-tree/schemaCreationUtilities.ts"],
```
Comment (Josmithr) [inline experimental/dds/tree2/src/test/tsconfig.json:12]: Probably needs to be reverted?

File: experimental/dds/tree2/src/class-tree/schemaCreationUtilities.ts (hunk 3 above)
Comment (Josmithr) [reply experimental/dds/tree2/src/class-tree/schemaCreationUtilities.ts:83]: Actually, looks like it isn't exported, so we can probably remove the `@alpha` tag

**Review by Josmithr — COMMENTED** [review]
Left a couple of comments that I think need to be addressed before merging, but the enum support looks great!

File: experimental/dds/tree2/src/class-tree/schemaCreationUtilities.ts (hunk 3 above)
Comment (CraigMacomber) [reply experimental/dds/tree2/src/class-tree/schemaCreationUtilities.ts:83]: The examples need it, so it should be exported as real users will need it as well (ex: to create a field that contains the enum in a schema)

File: experimental/dds/tree2/src/class-tree/schemaCreationUtilities.ts (hunk 5 above)
Comment (CraigMacomber) [reply experimental/dds/tree2/src/class-tree/schemaCreationUtilities.ts:161]: This is a less code duplication heavy version of enumFromString. I've updated the TODO to better indicate this.

File: experimental/dds/tree2/src/class-tree/schemaCreationUtilities.ts (hunk 2 above)
Comment (CraigMacomber) [reply experimental/dds/tree2/src/class-tree/schemaCreationUtilities.ts:63]: Added note to docs.

File: experimental/dds/tree2/src/class-tree/schemaCreationUtilities.ts (hunk 3 above)
Comment (CraigMacomber) [reply experimental/dds/tree2/src/class-tree/schemaCreationUtilities.ts:105]: Added examples using it to docs.

File: experimental/dds/tree2/src/class-tree/schemaCreationUtilities.ts (hunk 4 above)
Comment (CraigMacomber) [reply experimental/dds/tree2/src/class-tree/schemaCreationUtilities.ts]: done.

**Review by Josmithr — APPROVED** [review]
🚀

**Pull request merged**